jxl/src/container/mod.rs

```diff
@@ -0,0 +1,261 @@
+// Copyright (c) the JPEG XL Project Authors. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+//
+// Originally written for jxl-oxide.
+
+pub mod box_header;
+
+use box_header::*;
+
+use crate::error::Error;
+
+/// Container format parser.
+#[derive(Default)]
+pub struct ContainerParser {
+    state: DetectState,
+    buf: Vec<u8>,
+    codestream: Vec<u8>,
+    aux_boxes: Vec<(ContainerBoxType, Vec<u8>)>,
+    next_jxlp_index: u32,
+}
+
+impl std::fmt::Debug for ContainerParser {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("ContainerParser")
+            .field("state", &self.state)
+            .field("next_jxlp_index", &self.next_jxlp_index)
+            .finish_non_exhaustive()
+    }
+}
+
+#[derive(Debug, Default)]
+enum DetectState {
+    #[default]
+    WaitingSignature,
+    WaitingBoxHeader,
+    WaitingJxlpIndex(ContainerBoxHeader),
+    InAuxBox {
+        header: ContainerBoxHeader,
+        data: Vec<u8>,
+        bytes_left: Option<usize>,
+    },
+    InCodestream {
+        kind: BitstreamKind,
+        bytes_left: Option<usize>,
+    },
+    Done(BitstreamKind),
+}
+
+/// Structure of the decoded bitstream.
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum BitstreamKind {
+    /// Decoder can't determine structure of the bitstream.
+    Unknown,
+    /// Bitstream is a direct JPEG XL codestream without box structure.
+    BareCodestream,
+    /// Bitstream is a JPEG XL container with box structure.
+    Container,
+    /// Bitstream is not a valid JPEG XL image.
+    Invalid,
+}
+
+impl ContainerParser {
+    const CODESTREAM_SIG: [u8; 2] = [0xff, 0x0a];
+    const CONTAINER_SIG: [u8; 12] = [0, 0, 0, 0xc, b'J', b'X', b'L', b' ', 0xd, 0xa, 0x87, 0xa];
+
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn kind(&self) -> BitstreamKind {
+        match self.state {
+            DetectState::WaitingSignature => BitstreamKind::Unknown,
+            DetectState::WaitingBoxHeader
+            | DetectState::WaitingJxlpIndex(..)
+            | DetectState::InAuxBox { .. } => BitstreamKind::Container,
+            DetectState::InCodestream { kind, .. } | DetectState::Done(kind) => kind,
+        }
+    }
+
+    pub fn feed_bytes(&mut self, input: &[u8]) -> Result<(), Error> {
+        let state = &mut self.state;
+        let buf = &mut self.buf;
+        buf.extend_from_slice(input);
```

Comment:
Here and below, I would use buf.try_reserve() to make sure that we do not get panics due to failed allocations

Reply:
Also addressed in c3e8a51.